File: apps/mobile/androidApp/src/main/kotlin/co/heva/piro/android/push/PushPayloadUnsealer.kt
package co.heva.piro.android.push

import android.content.Context
import android.util.Base64
import org.json.JSONObject
import java.security.PrivateKey
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/** What the backend sealed for this device. Mirrors the fields the cleartext data map used to carry. */
data class PushPayload(
    val title: String,
    val body: String,
    val eventKey: String,
    val alertId: Int,
    val url: String?,
)

/**
 * Opens an envelope sealed by the backend's PushPayloadSealer.
 *
 * The scheme, which must stay byte-for-byte in step with the server:
 *  - ECDH P-256 between this device's private key and the ephemeral public key carried in the envelope
 *  - HKDF-SHA256 over the shared secret, salted with that ephemeral public key, info "piro-push-v1"
 *  - AES-256-GCM, 12-byte nonce, 16-byte tag appended to the ciphertext, AAD "piro-push-v1"
 *
 * The ephemeral public key travels in the clear on purpose: it is not a secret, and without this
 * device's private key it yields nothing. That is what lets the relay forward a payload it cannot read.
 */
object PushPayloadUnsealer {

    private const val VERSION = 1
    private const val AAD = "piro-push-v1"
    private const val HKDF_INFO = "piro-push-v1"
    private const val NONCE_BYTES = 12
    private const val TAG_BITS = 128

    /**
     * Decrypts [envelopeBase64Url] and parses the payload.
     *
     * @throws IllegalArgumentException if the envelope is malformed or its version is unsupported.
     * @throws javax.crypto.AEADBadTagException if it was not sealed for this device.
     */
    fun unseal(context: Context, envelopeBase64Url: String): PushPayload =
        unseal(PushKeyStore.privateKey(context), envelopeBase64Url)

    /** Testable overload taking the key directly, so unit tests need no Android context. */
    internal fun unseal(privateKey: PrivateKey, envelopeBase64Url: String): PushPayload {
        val envelope = JSONObject(String(decodeBase64Url(envelopeBase64Url), Charsets.UTF_8))

        val version = envelope.getInt("V")
        require(version == VERSION) { "Unsupported push envelope version $version" }

        val ephemeralRaw = decodeBase64Url(envelope.getString("Epk"))
        val nonce = decodeBase64Url(envelope.getString("N"))
        val sealed = decodeBase64Url(envelope.getString("Ct"))

        require(nonce.size == NONCE_BYTES) { "Nonce must be $NONCE_BYTES bytes, got ${nonce.size}" }
        require(sealed.size > TAG_BITS / 8) { "Ciphertext is too short to carry a tag" }

        val agreement = KeyAgreement.getInstance("ECDH").apply {
            init(privateKey)
            doPhase(PushKeyStore.publicKeyFromRawPoint(ephemeralRaw), true)
        }
        val shared = agreement.generateSecret()

        val key = hkdfSha256(
            ikm = shared,
            salt = ephemeralRaw,
            info = HKDF_INFO.toByteArray(Charsets.US_ASCII),
            length = 32,
        )

        val plaintext = try {
            Cipher.getInstance("AES/GCM/NoPadding").run {
                init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_BITS, nonce))
                updateAAD(AAD.toByteArray(Charsets.US_ASCII))
                doFinal(sealed)
            }
        } finally {
            key.fill(0)
            shared.fill(0)
        }

        return parse(JSONObject(String(plaintext, Charsets.UTF_8)))
    }

    private fun parse(json: JSONObject) = PushPayload(
        title = json.optString("title", ""),
        body = json.optString("body", ""),
        eventKey = json.optString("eventKey", ""),
        alertId = json.optInt("alertId", 0),
        // optString returns "null" for a JSON null, which would render as literal text downstream.
        url = json.optString("url", "").takeIf { it.isNotEmpty() && it != "null" },
    )

    /**
     * HKDF-SHA256 (RFC 5869). Only one output block is ever needed here (32 bytes = one SHA-256 block),
     * so the expand step is a single iteration rather than the general loop.
     */
    private fun hkdfSha256(ikm: ByteArray, salt: ByteArray, info: ByteArray, length: Int): ByteArray {
        require(length <= 32) { "This HKDF only emits a single SHA-256 block" }
        val mac = Mac.getInstance("HmacSHA256")

        mac.init(SecretKeySpec(salt, "HmacSHA256"))
        val pseudoRandomKey = mac.doFinal(ikm)

        mac.init(SecretKeySpec(pseudoRandomKey, "HmacSHA256"))
        mac.update(info)
        mac.update(1.toByte())
        return mac.doFinal().copyOf(length)
    }

    private fun decodeBase64Url(value: String): ByteArray =
        Base64.decode(value, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
}

File: apps/mobile/iosApp/Piro/Push/PushPayloadUnsealer.swift
import CryptoKit
import Foundation

/// What the backend sealed for this device.
struct PushPayload: Equatable {
    let title: String
    let body: String
    let eventKey: String
    let alertId: Int
    let url: String?
}

enum PushUnsealError: Error, Equatable {
    case notBase64Url
    case malformedEnvelope
    case unsupportedVersion(Int)
    case badNonceLength(Int)
    case ciphertextTooShort
    case noPrivateKey
    case decryptionFailed
}

/// Opens an envelope sealed by the backend's `PushPayloadSealer` (RFC 0017).
///
/// The scheme, which must stay byte-for-byte in step with the server and the Android client:
///  - ECDH P-256 between this device's private key and the ephemeral public key in the envelope
///  - HKDF-SHA256 over the shared secret, salted with that ephemeral public key, info `piro-push-v1`
///  - AES-256-GCM, 12-byte nonce, 16-byte tag appended to the ciphertext, AAD `piro-push-v1`
///
/// The ephemeral public key travels in the clear on purpose: it is not a secret, and without this
/// device's private key it yields nothing. That is what lets a relay forward a payload it cannot read.
///
/// Verified against an envelope produced by the real server sealer — the interop that matters here is
/// with `PushPayloadSealer`, not with a Swift-side round trip, since a self-consistent bug would pass
/// the latter. There is no iOS test target yet to keep that check running; this vector reproduces it:
///
///     private key (base64url raw): 6Z--VmFjwVhHk-aUJz1zi-CN8eOC2IWv4tznQuzR_3U
///     decrypts to: title "Critical alert", body "heva-api is down",
///                  eventKey "alert.raised", alertId 4242, url "piro://alert/4242"
enum PushPayloadUnsealer {

    private static let version = 1
    private static let associatedData = Data("piro-push-v1".utf8)
    private static let hkdfInfo = Data("piro-push-v1".utf8)
    private static let nonceLength = 12
    private static let tagLength = 16

    /// Decrypts an envelope using the key held in the Keychain.
    static func unseal(_ envelopeBase64Url: String) throws -> PushPayload {
        guard let key = PushKeyStore.privateKey() else { throw PushUnsealError.noPrivateKey }
        return try unseal(envelopeBase64Url, with: key)
    }

    /// Testable overload taking the key directly.
    static func unseal(
        _ envelopeBase64Url: String,
        with privateKey: P256.KeyAgreement.PrivateKey
    ) throws -> PushPayload {
        guard let envelopeData = PushKeyStore.base64UrlDecode(envelopeBase64Url) else {
            throw PushUnsealError.notBase64Url
        }
        guard let envelope = try? JSONSerialization.jsonObject(with: envelopeData) as? [String: Any] else {
            throw PushUnsealError.malformedEnvelope
        }

        // The version is checked before anything else and is also bound into the AAD, so a downgrade
        // fails to decrypt rather than being silently reinterpreted.
        guard let v = envelope["V"] as? Int else { throw PushUnsealError.malformedEnvelope }
        guard v == version else { throw PushUnsealError.unsupportedVersion(v) }

        guard let epk = envelope["Epk"] as? String,
              let n = envelope["N"] as? String,
              let ct = envelope["Ct"] as? String,
              let ephemeralRaw = PushKeyStore.base64UrlDecode(epk),
              let nonceData = PushKeyStore.base64UrlDecode(n),
              let sealedData = PushKeyStore.base64UrlDecode(ct)
        else { throw PushUnsealError.malformedEnvelope }

        guard nonceData.count == nonceLength else {
            throw PushUnsealError.badNonceLength(nonceData.count)
        }
        guard sealedData.count > tagLength else { throw PushUnsealError.ciphertextTooShort }

        guard let ephemeralPublic = try? P256.KeyAgreement.PublicKey(x963Representation: ephemeralRaw) else {
            throw PushUnsealError.malformedEnvelope
        }

        guard let shared = try? privateKey.sharedSecretFromKeyAgreement(with: ephemeralPublic) else {
            throw PushUnsealError.decryptionFailed
        }

        // Salted with the ephemeral public key so two pushes to the same device never derive the same
        // content key. CryptoKit zeroes the derived material itself when it goes out of scope.
        let key = shared.hkdfDerivedSymmetricKey(
            using: SHA256.self,
            salt: ephemeralRaw,
            sharedInfo: hkdfInfo,
            outputByteCount: 32)

        // The server appends the tag to the ciphertext, which is exactly what `combined` expects once
        // the nonce is prefixed.
        guard let nonce = try? AES.GCM.Nonce(data: nonceData),
              let box = try? AES.GCM.SealedBox(
                  combined: Data(nonce) + sealedData),
              let plaintext = try? AES.GCM.open(box, using: key, authenticating: associatedData)
        else { throw PushUnsealError.decryptionFailed }

        guard let json = try? JSONSerialization.jsonObject(with: plaintext) as? [String: Any] else {
            throw PushUnsealError.malformedEnvelope
        }

        return PushPayload(
            title: json["title"] as? String ?? "",
            body: json["body"] as? String ?? "",
            eventKey: json["eventKey"] as? String ?? "",
            alertId: json["alertId"] as? Int ?? 0,
            // A JSON null decodes to NSNull, which would render as "<null>" if passed through.
            url: (json["url"] as? String).flatMap { $0.isEmpty ? nil : $0 })
    }
}
